migrations/entitlements/migration.go

```diff
@@ -0,0 +1,240 @@
+							Entitlements: supportedEntitlements,
+						}
+					}
+					t.Authorization = interpreter.ConvertSemaAccessToStaticAuthorization(inter, auth)
+				}
+			}
+		case *interpreter.CapabilityStaticType:
+			convertLegacyStaticType(t.BorrowType)
+		case *interpreter.VariableSizedStaticType:
+			convertLegacyStaticType(t.Type)
+		case *interpreter.ConstantSizedStaticType:
+			convertLegacyStaticType(t.Type)
+		case *interpreter.DictionaryStaticType:
+			convertLegacyStaticType(t.KeyType)
+			convertLegacyStaticType(t.ValueType)
+		case *interpreter.OptionalStaticType:
+			convertLegacyStaticType(t.Type)
+		}
+	}
+
+	convertLegacyStaticType(staticType)
+	semaType := inter.MustConvertStaticToSemaType(staticType)
+	entitledType := ConvertToEntitledType(semaType)
+
+	switch v := v.(type) {
+	case *interpreter.EphemeralReferenceValue:
+		entitledReferenceType := entitledType.(*sema.ReferenceType)
+		staticAuthorization := interpreter.ConvertSemaAccessToStaticAuthorization(inter, entitledReferenceType.Authorization)
+		convertedValue := ConvertValueToEntitlements(inter, v.Value)
+		// if the underlying value did not change, we still want to use the old value in the newly created reference
+		if convertedValue == nil {
+			convertedValue = v.Value
+		}
+		return interpreter.NewEphemeralReferenceValue(
+			inter,
+			staticAuthorization,
+			convertedValue,
+			entitledReferenceType.Type,
+			interpreter.EmptyLocationRange,
+		)
+
+	case *interpreter.StorageReferenceValue:
+		// a stored value will in itself be migrated at another point, so no need to do anything here other than change the type
+		entitledReferenceType := entitledType.(*sema.ReferenceType)
+		staticAuthorization := interpreter.ConvertSemaAccessToStaticAuthorization(inter, entitledReferenceType.Authorization)
+		return interpreter.NewStorageReferenceValue(
+			inter,
+			staticAuthorization,
+			v.TargetStorageAddress,
+			v.TargetPath,
+			entitledReferenceType.Type,
+		)
+
+	case *interpreter.ArrayValue:
+		entitledArrayType := entitledType.(sema.ArrayType)
+		arrayStaticType := interpreter.ConvertSemaArrayTypeToStaticArrayType(inter, entitledArrayType)
+
+		iterator := v.Iterator(inter)
+
+		newArray := interpreter.NewArrayValueWithIterator(inter, arrayStaticType, v.GetOwner(), uint64(v.Count()), func() interpreter.Value {
+			return iterator.Next(inter)
+		})
+		return newArray
+
+	case *interpreter.DictionaryValue:
+		entitledDictionaryType := entitledType.(*sema.DictionaryType)
+		dictionaryStaticType := interpreter.ConvertSemaDictionaryTypeToStaticDictionaryType(inter, entitledDictionaryType)
+
+		var values []interpreter.Value
+
+		v.Iterate(inter, func(key, value interpreter.Value) (resume bool) {
+			values = append(values, key)
+			values = append(values, value)
+			return true
+		})
+
+		newDict := interpreter.NewDictionaryValue(
+			inter,
+			interpreter.EmptyLocationRange,
+			dictionaryStaticType,
+			values...,
+		)
+		return newDict
```

Comment:
not sure, maybe transferring the value and changing the static-type could be sufficient as well (could avoid having to manually iterate and copy).
e.g:
```go
transferredDictionary := v.Transfer(
	interpreter,
	EmptyLocationRange,
	atree.Address{},
	true,
	nil,
	nil,
).(*DictionaryValue)

transferredDictionary.Type = dictionaryStaticType
transferredDictionary.semaType = entitledDictionaryType
return transferredDictionary
```

Reply:
Tried this and ran into the same crash :/

Reply:
What is the crash?

Reply:
Previously this was crashing in the atree code, but it seems like with the changes to the migration framework since then this no longer crashes. However this does still fail tests; the `MigrateDictOfValues` test specifically loads the dictionary post-migration and the static type of the dict is not properly converted using this code, whereas creating a new dictionary succeeds.

Reply:
Interesting. I guess Cadence (atree?) does not currently support changes to the container's static type. 

For now, recreating the container works and is probably sufficient, but we might run into problems during the migration of large real-world containers, and might need to go back and address this overhead.